Magento 2 cron needs zombies to exist

kierenevans opened the pull request:
1. Main cron process spawns new standalone processes as children
2. Main cron process exits after doing it's tasks
3. Zombies created out of the forked processes
4. PID 1 (tini) becomes parent of zombies
5. PID 1 (tini) reaps the zombies
6. Standalone processes never complete

andytson-inviqa commented:
if a bash shell for the cron task could run the task then `wait` (waits for all children to finish) then maybe no need

kierenevans commented:
I think that's what we did in https://github.com/continuouspipe/dockerfiles/pull/180/files with supervisor PID 1, will adjust.

andytson-inviqa commented:
I think that's just for allowing shell and not just a single command

more like (maybe) `bash -c '... &  && wait'`. I don't know if the cron.php's forks become the bash process's children tho

kierenevans committed:
Magento 2 Cron needs Zombies to exist

1. Main cron process spawns new standalone processes as children
2. Main cron process exits after doing it's tasks
3. Zombies created out of the forked processes
4. PID 1 (tini) becomes parent of zombies
5. PID 1 (tini) reaps the zombies
6. Standalone processes do not complete

kierenevans commented:
Unfortunately they don't: https://github.com/magento/magento2/blob/2.4.0/app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php#L255-L261

kierenevans commented:
We'll try this where we've been seeing issues :crossed_fingers:

andytson-inviqa commented:
that's spawned children rather than forks btw

kierenevans commented on src/magento2/harness.yml:
```diff
@@ -153,7 +153,7 @@ attributes:
     cron:
       jobs:
-        - "= (@('app.mode') == 'production' ? '* * * * * /cron-run-with-env.sh /app/bin/magento cron:run' : '')"
+        - "= (@('app.mode') == 'production' ? '* * * * * /cron-run-with-env.sh /bin/bash -c \"/app/bin/magento cron:run & && wait\"' : '')"
```
```
bash: -c: line 0: syntax error near unexpected token `&&'
bash: -c: line 0: `/bin/bash -c /app/bin/magento cron:run & && wait'
```

kierenevans replied:
```suggestion
        - "= (@('app.mode') == 'production' ? '* * * * * /cron-run-with-env.sh /bin/bash -m -c \\'\"/app/bin/magento cron:run & wait\"\\'' : '')"
```

kierenevans committed:
Enable job control

kierenevans committed:
No need for job control
